eth/fetcher/fetcher.go

```diff
@@ -147,21 +151,23 @@ type Fetcher struct {
 // New creates a block fetcher to retrieve blocks based on hash announcements.
 func New(getBlock blockRetrievalFn, verifyHeader headerVerifierFn, broadcastBlock blockBroadcasterFn, chainHeight chainHeightFn, insertChain chainInsertFn, dropPeer peerDropFn) *Fetcher {
 	return &Fetcher{
-		notify:         make(chan *announce),
-		inject:         make(chan *inject),
-		headerFilter:   make(chan chan *headerFilterTask),
-		bodyFilter:     make(chan chan *bodyFilterTask),
-		done:           make(chan common.Hash),
-		quit:           make(chan struct{}),
-		announces:      make(map[string]int),
-		announced:      make(map[common.Hash][]*announce),
-		fetching:       make(map[common.Hash]*announce),
-		fetched:        make(map[common.Hash][]*announce),
-		completing:     make(map[common.Hash]*announce),
-		queue:          prque.New(nil),
-		queues:         make(map[string]int),
-		queued:         make(map[common.Hash]*inject),
-		getBlock:       getBlock,
+		notify:       make(chan *announce),
+		inject:       make(chan *inject),
+		headerFilter: make(chan chan *headerFilterTask),
+		bodyFilter:   make(chan chan *bodyFilterTask),
+		done:         make(chan common.Hash),
+		quit:         make(chan struct{}),
+		announces:    make(map[string]int),
+		announced:    make(map[common.Hash][]*announce),
+		fetching:     make(map[common.Hash]*announce),
+		fetched:      make(map[common.Hash][]*announce),
+		completing:   make(map[common.Hash]*announce),
+		queue:        prque.New(nil),
+		queues:       make(map[string]int),
+		queued:       make(map[common.Hash]*inject),
+		getBlock:     getBlock,
+		// This can be replaced this with a more optimized lookup
+		hasBlock:       func(hash common.Hash) bool { return getBlock(hash) != nil },
```

Comment:
We can probably replace this with a blockchain.HasBlock or something that's actually not dependent on so many lookups